dra-evolution/pkg/api/capacity_types.go

```diff
@@ -3,168 +3,134 @@ package api
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/types"
 )
 
-const (
-	DevMgmtAPIVersion = "devmgmtproto.k8s.io/v1alpha1"
-)
-
-// DevicePool represents a collection of devices managed by a given driver. How
+// ResourcePool represents a collection of devices managed by a given driver. How
 // devices are divided into pools is driver-specific, but typically the
 // expectation would a be a pool per identical collection of devices, per node.
 // It is fine to have more than one pool for a given node, for the same driver.
-type DevicePool struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
+//
+// Where a device gets published may change over time. The unique identifier
+// for a device is the tuple `<driver name>/<node name>/<device name>`. Each
+// of these names is a DNS label or domain, so it is okay to concatenate them
+// like this in a string with a slash as separator.
+//
+// Consumers should be prepared to handle situations where the same device is
+// listed in different pools, for example because the producer already added it
+// to a new pool before removing it from an old one. Should this occur, then
+// there is still only one such device instance.
+type ResourcePool struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-	Spec   DevicePoolSpec   `json:"spec,omitempty"`
-	Status DevicePoolStatus `json:"status,omitempty"`
-}
+	Spec ResourcePoolSpec `json:"spec"`
 
-// DevicePoolSpec identifies the driver and contains the data for the pool
-// prior to any allocations.
-// NOTE: It's not clear that spec/status is the right model for this data.
-type DevicePoolSpec struct {
-	// NodeName is the name of the node containing the devices in the pool.
-	// For network attached devices, this may be empty.
-	// +optional
-	NodeName *string `json:"nodeName,omitempty"`
+	// Stretch goal for 1.31: define status
+	//
+	// To discuss:
+	// - Who writes that status?
+	//   After https://github.com/kubernetes/kubernetes/pull/125163 (implements
+	//   https://github.com/kubernetes/enhancements/pull/4667), kubelet is not
+	//   involved with publishing ResourcePools, they come directly from the driver.
+	// - What information should be in it?
+	// - Does it need to be a sub-resource? This would make it harder
+	//   for a driver to publish a new device (spec) and its corresponding
+	//   health information (status).
+}
 
-	// Driver is the name of the DeviceDriver that created this object and
-	// owns the data in it.
-	// +required
-	Driver string `json:"driver,omitempty"`
+type ResourcePoolSpec struct {
+	// NodeName identifies the node which provides the devices. All devices
+	// are local to that node.
+	//
+	// This is currently required, but this might get relaxed in the future.
+	NodeName string `json:"nodeName"`
```

Comment:
Candidate for list-of-one-of pattern because we might add a NodeSelector or just make some resources available from anywhere.  IIRC the reason not to just use a node selector is because we want on-node drivers to be authorized only for their own node's pools, right?  I'm not well versed in the authorizer subsystem, but wouldn't it be possible to look at a node selector and if it exactly matches `matchFields: metadata.name` AND driverName matches the credential, then allow?

Assuming that's naive, I will analyze this anyway :)

We know that, in general, control-planes are updated before nodes, and I presume that includes drivers, but on-node drivers are exactly the ones who WON'T need a new option here. 

Suppose we have a driver for some rack-switched thing (e.g. a PCIe switch).  We install it before updating the cluster.  It wants to set `nodeSelector: example.com/rack == r93` on this pool, but the apiserver isn't updated.  We'll discard that information, and the apiserver will barf because `nodeName` was not set.  Good.
 
The other direction is the risk - a pool is served which sets NodeSelector and not NodeName.

It seems like a safe assumption that kube-scheduler gets updated in conjunction with the control plane, so I am not super worried about that.  But we know we have schedulers and autoscalers out of the core of k8s.  If there were a back-rev scheduler which saw such an object, it would know that something is wrong because NodeName was not populated.  Good.

As long as we never change `nodeName: ""` to mean "all nodes", it is not ambiguous.  So how would a driver indicate all nodes?  `nodeSelector: NotIn []` or something, I guess?  Or we could make this `*string` and say `""` means "all".  Or we could add a `Scope: Node | Nodes | Cluster` field.

So maybe we don't need special handling here.

Reply:
> I'm not well versed in the authorizer subsystem, but wouldn't it be possible to look at a node selector and if it exactly matches matchFields: metadata.name AND driverName matches the credential, then allow?

I doubt that this would work. Keeping the explicit field is definitely simpler.

> So maybe we don't need special handling here.

I came to the same conclusion. The only slightly icky part is that right now, `NodeName` will never be empty and the apiserver will validate that. But at the same time we know that this may change, so current clients should not rely on that. That's the reason for the "might get relaxed in the future" comment. What they can do is `pool.spec.nodeName == node.name`. This will continue to work, even if the field becomes empty at some point.

A one-of struct would send a stronger signal to client developers that they have to be prepared for some future change. But that's not needed here.

Reply:
> This will continue to work, even if the field becomes empty at some point.

"Work" in the sense that a scheduler is not doing something wrong. To allocate a network-attached device, an old scheduler has to be extended to also consider a future node selector.

Reply:
I think we agree - an inline one-of is sufficiently future safe.  If you REALLY want to make it a named struct, OK, but only if it actually adds readability.

e.g.

```
spec:
     nodeName: foobar
```

vs

```
spec:
  availableFrom:
    nodeName: foobar
```